numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1933,6 +1937,367 @@ operand_different_than_broadcast: {
+     */
+    npy_intp sizeof_axisdata = NIT_AXISDATA_SIZEOF(itflags, ndim, nop);
+    NpyIter_AxisData *axisdata = NIT_AXISDATA(iter);
+    npyiter_opitflags *op_itflags = NIT_OPITFLAGS(iter);
+
+    /*
+     * We can only continue as long as we are within the maximum allowed size.
+     * When no buffering is needed and GROWINNER is set, we don't have to
+     * worry about this maximum.
+     */
+    npy_intp maximum_size = NBF_BUFFERSIZE(bufferdata);
+
+    /* The cost factor defined by: (1 + n_buffered) */
+    int cost = 1;
+
+    for (int iop = 0; iop < nop; ++iop) {
+        op_single_stride_dims[iop] = 1;
+        op_reduce_outer_dim[iop] = 0;
+        if (op_itflags[iop] & NPY_OP_ITFLAG_CAST) {
+            cost += 1;
+        }
+    }
+
+    /*
+     * Once a reduce operand reaches a ==0/!=0 stride flip, this dimension
+     * becomes the outer reduce dimension.
+     * We may continue growing, but only if strides align for any such operand.
+     */
+    int outer_reduce_dim = 0;
+
+    npy_intp size = axisdata->shape;  /* the current total size */
+
+    int best_dim = 0;
+    int best_cost = cost;
+    /* The size of the "outer" iteration and all previous dimensions: */
+    npy_intp best_size = size;
+    npy_intp best_coresize = 1;
+
+    NPY_IT_DBG_PRINT("Iterator: discovering best core size\n");
+    for (int idim = 1; idim < ndim; idim++) {
+        if (outer_reduce_dim) {
+            /* Cannot currently expand beyond reduce dim! */
+            break;
+        }
+        if (size >= maximum_size &&
+                (cost > 1 || !(itflags & NPY_ITFLAG_GROWINNER))) {
+            /* Exceeded buffer size, can only improve without buffers and growinner. */
+            break;
+        }
+
+        npy_intp *prev_strides = NAD_STRIDES(axisdata);
+        npy_intp prev_shape = NAD_SHAPE(axisdata);
+        NIT_ADVANCE_AXISDATA(axisdata, 1);
+        npy_intp *strides = NAD_STRIDES(axisdata);
+
+        int iop;
+        for (iop = 0; iop < nop; iop++) {
+            /* Check that we set things up nicely (if shape is ever 1) */
+            assert((axisdata->shape == 1) ? (prev_strides[iop] == strides[iop]) : 1);
+            /*
+             * Best case: the strides collaps for this operand, all fine.
+             * Keep track at which single-stride or outer dims we are.
+             */
+            if (prev_strides[iop] * prev_shape == strides[iop]) {
+                if (op_single_stride_dims[iop] == idim) {
+                    op_single_stride_dims[iop] += 1;
+                }
+                continue;
+            }
+
+            if (op_single_stride_dims[iop] == idim) {
+                /*
+                 * Operand now requires buffering (if it was not already).
+                 * NOTE: This is technically not true since we may still use
+                 *       an outer reduce at this point.
+                 *       So it prefers a non-reduce setup, which seems not
+                 *       ideal, but OK.
```

Comment:
this might work, but not sure if it should matter right now (should make an issue probably that this all should be tweaked)...

The point is, if we use a "reduce" setup (i.e. two strides), then we can do one more dimension without buffering.

And maybe/probably it works to change this if and move it to the beginning (before the check for identical strides).
(I have to think about it myself.)

Reply:
Talking with Matti today so hopefully we can push it through (with some notes in an issue).  So a bit of a note to self:  Changing it here is simple, I think.  The difference will be that `using_reduce` needs to be discovered below by re-checking operands.
Reduce-loops are unfavorable since you need to call the inner-loop/`iternext` more often (but, you save on buffer filling).  That trade-off may already be covered by preferring larger cores, but not sure.

This whole "finding the best" is very crude and I am sure it can be improved.  Just also don't want to overcomplicate it and make it slow.

Reply:
Let's leave that for a further investigation/PR. Maybe we can open a follow up issue to track more cleanup and optimizations.